operator-sdk: run bundle{-upgrade} support insecure registry server

avalluri opened the pull request:
Supporting insecure registry is useful for testing operator installation
from a bundle image stored at the local registry.

Closes #4815

**Description of the change:**

Add a new boolean command-line option(defaults to false) named `--allow-insecure` for both `operator-sdk run bundle` and `operator-sdk run bundle-upgrade`. When enabled this argument the commands ignores the SSL errors that occurred while communicating the docker registry. This allows pulling the bundle images from insecure(usually local) docker registries.

**Motivation for the change:**

Using a local insecure registry is a possible case for storing the generated bundle image, and test the operator installation.

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [x] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

estroz commented on internal/olm/operator/bundle/install.go:
```diff
@@ -51,6 +51,7 @@ func (i *Install) BindFlags(fs *pflag.FlagSet) {
 
 	// --mode is hidden so only users who know what they're doing can alter add mode.
 	fs.StringVar((*string)(&i.BundleAddMode), "mode", "", "mode to use for adding bundle to index")
+	fs.BoolVar(&i.IndexImageCatalogCreator.AllowInsecure, "allow-insecure", false, "allows pulling bundle image from an insecure registry server")
```
How about mirroring `opm` and naming this
```suggestion
	fs.BoolVar(&i.IndexImageCatalogCreator.SkipTLS, "skip-tls", false, "skip authentication of image registry TLS certificate when pulling a bundle image in-cluster")
```

avalluri replied:
> How about mirroring `opm` and naming this

Sounds good, i will update the PR.

avalluri replied:
Renamed the option as suggested. Also moved the flag to `IndexImageCatalogCreator` in `internal/olm/operator/registry/index_image.go` so the option will be available to both `run bundle` and `run bundle-upgrade`.

estroz commented on internal/olm/operator/registry/index/registry_pod.go:
```diff
@@ -50,6 +50,8 @@ type BundleItem struct {
 	// AddMode describes how the bundle should be added to an index image.
 	AddMode BundleAddMode `json:"mode"`
+	// SkipTLS controls wether to ignore SSL errors while pulling bundle image from registry server.
+	SkipTLS bool `json:"SkipTLS"`
```
This field should be added to `RegistryPod` not `BundleItem` since it's globally configured and not on a per-bundle basis.

jberkhahn replied:
that's also where the preexisting cert stuff is anyway :+1:

jmrodri replied:
agreed

avalluri replied:
Done. Changed as @estroz suggested.

Venkat19967 commented:
/lgtm

jmrodri commented on internal/olm/operator/registry/index/registry_pod.go:
```diff
@@ -302,7 +304,7 @@ func newBool(b bool) *bool {
 const cmdTemplate = `/bin/mkdir -p {{ dirname .DBPath }} && \
 {{- range $i, $item := .BundleItems }}
-/bin/opm registry add -d {{ $.DBPath }} -b {{ $item.ImageTag }} --mode={{ $item.AddMode }}{{ if $.CASecretName }} --ca-file=/certs/cert.pem{{ end }} && \
+/bin/opm registry add -d {{ $.DBPath }} -b {{ $item.ImageTag }} --mode={{ $item.AddMode }}{{ if $.CASecretName }} --ca-file=/certs/cert.pem{{ end }} --skip-tls={{ $item.SkipTLS }} && \
```
:+1:  Great catch. When I was playing around to add this locally I didn't notice having to do this.

jmrodri commented on internal/olm/operator/registry/index/registry_pod_test.go:
```diff
@@ -114,6 +132,11 @@ var _ = Describe("RegistryPod", func() {
+					BundleItem{
+						ImageTag: "localhost/example-operator-bundle:1.0.1",
+						AddMode:  SemverBundleAddMode,
+						SkipTLS:  true,
```
Like was said earlier, this field should be on RegistryPod.

avalluri replied:
Done.

jmrodri commented on internal/olm/operator/registry/index/registry_pod_test.go:
```diff
@@ -91,6 +92,14 @@ var _ = Describe("RegistryPod", func() {
 			})
 
+			It("should create a registry with custom database path", func() {
+				customDBPath := "/foo/bar/database.db"
```
Do we allow a custom database? This seems like we're mixing 2 different items.

avalluri replied:
> Do we allow a custom database?

I am not sure, in case, we don't support then we could remove `dbPath` parameter from `containerCommandFor` in the unit tests.

avalluri replied:
At the same time, I don't see any logical reason for not supporting the custom database path.

jmrodri commented on internal/olm/operator/registry/index/registry_pod_test.go:
```diff
@@ -224,7 +247,7 @@ func containerCommandFor(dbPath string, items []BundleItem, hasCA bool) string {
+		additions.WriteString(fmt.Sprintf("/bin/opm registry add -d %s -b %s --mode=%s%s --skip-tls=%v && \\\n", dbPath, item.ImageTag, item.AddMode, caFlag, item.SkipTLS))
 	}
-	return fmt.Sprintf("/bin/mkdir -p /database && \\\n%s/bin/opm registry serve -d /database/index.db -p 50051\n", additions.String())
+	return fmt.Sprintf("/bin/mkdir -p %s && \\\n%s/bin/opm registry serve -d %s -p 50051\n", filepath.Dir(dbPath), additions.String(), dbPath)
```
I didn't think we were allowing custom database path. I think that should be a separate PR.

avalluri replied:
I added this custom database change in this PR as the CI reported a failure:

```
go vet ./...
tools/scripts/fetch golangci-lint 1.31.0 && tools/bin/golangci-lint run
golangci-lint missing or not version '1.31.0', downloading...
golangci/golangci-lint info checking GitHub for tag 'v1.31.0'
golangci/golangci-lint info found version: 1.31.0 for v1.31.0/linux/amd64
golangci/golangci-lint info installed /home/runner/work/operator-sdk/operator-sdk/tools/bin/golangci-lint
internal/olm/operator/registry/index/registry_pod_test.go:234:26: `containerCommandFor` - `dbPath` always receives `defaultDBPath` (`"/database/index.db"`) (unparam)
func containerCommandFor(dbPath string, items []BundleItem, hasCA bool) string {
                         ^
Makefile:122: recipe for target 'test-sanity' failed
make: *** [test-sanity] Error 1
Error: Process completed with exit code 2.
```

jmrodri reviewed (changes requested):
All in all a GREAT PR. I have 2 items:
1. the custom DB path should be a separate PR
2. the SkipTLS field should be on RegistryPod like others mentioned.

avalluri commented:
> All in all a GREAT PR.

Thanks! for your review time. Sorry for late responce.
 I have 2 items:

> 
> 1. the custom DB path should be a separate PR

I removed this change from this PR. I will try to send it as a separate PR.

> 2. the SkipTLS field should be on RegistryPod like others mentioned.

Changed as suggested.

estroz reviewed (approved):
/lgtm

jmrodri reviewed (approved):
/lgtm
